MapLoader with TTL

mustafaiman opened the pull request:
Adds EntryStore(MapStore with expiration times support).
Introduces interfaces; `EntryStore` and `EntryLoader`.
Introduces `Clock#toSystemCurrentTimeMillis` method.

vojtechtoman commented on hazelcast/src/main/java/com/hazelcast/map/impl/operation/PutFromLoadAllBackupOperation.java:
```diff
@@ -34,33 +34,36 @@
             if (includesExpirationTime) {
-                i++;
+                long expirationTime = (long) mapServiceContext.toObject(keyValueSequence.get(i++));
+                final Object object = mapServiceContext.toObject(value);
```
Minor: this line (and L64) can be moved outside the if statement

vojtechtoman commented on hazelcast/src/main/java/com/hazelcast/map/EntryStore.java:
```diff
@@ -38,5 +38,5 @@
  * @param <V> type of the EntryStore value
  */
-public interface EntryStore<K, V> extends EntryLoader<K, V>, MapStore<K, EntryLoaderEntry<V>> {
+public interface EntryStore<K, V> extends EntryLoader<K, V>, MapStore<K, ExtendedValue<V>> {
```
Why does `EntryStore` extend `EntryLoader`? For convenience?

mustafaiman replied:
EntryStore extends EntryLoader just as MapStore extends MapLoader. That is because even for plain map operations, MapStore/EntryStore needs to be able to retrieve old values from storage(to return old values). Also other operations such as remove-if-same or replace require loading values from underlying store.

vojtechtoman replied:
Thanks for the clarification, somehow I missed that `MapStore` extends `MapLoader` as well.

vojtechtoman commented on hazelcast/src/main/java/com/hazelcast/map/ExtendedValue.java:
```diff
@@ -19,12 +19,11 @@
  * @param <V> the type of the value object
  */
-public class EntryLoaderEntry<V> {
+public class ExtendedValue<V> {
```
The name `ExtendedValue` feels a bit too generic to me. Without further context, it it unclear what its purpose is, and combined with the fact that it sits directly in `com.hazelcast.map`, it may look more important that it is. The Javadoc clarifies that but still... Since this class it tightly bound to `MapLoader`/`MapStore`, what about making it a nested class of the `MapLoader` interface, or at least change its name to something like MapLoaderValue`? (Having said that, I know you played with different ideas in older commits.)

ahmetmircik replied:
Movement to map package is handled here: https://github.com/hazelcast/hazelcast/pull/15149

ahmetmircik replied:
alternative name suggestion: 
since we use ExtendedValue in both store and load mechanism, `MetaDataAwareValue` can be another alternative.

mustafaiman replied:
I appreciate the ideas about naming. I was actually going to go with `MapLoaderValue` but I did not because of the reason Ahmet stated. I think `MetadataAware` value sounds good. Plus, we might include more metadata in it in the future if requested. I will also include it as a nested class inside NOT `MapLoader` but `EntryLoader` as I think @vojtechtoman meant to suggest.

vojtechtoman commented on hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java:
```diff
@@ -343,6 +356,10 @@ public Record loadRecordOrNull(Data key, boolean backup, Address callerAddress)
     }
 
+    private long expirationTimeToTtl(long definedExpirationTime) {
+        return definedExpirationTime - System.currentTimeMillis();
```
Shouldn't this use custom `Clock`, too, like `AbstractMapDataStore.getUserExpirationTime()`?

mustafaiman replied:
We convert user expiration time to JVM based epoch as soon as we get them. Here `definedExpirationTime` is already an offset from JVM epoch.

vojtechtoman commented on hazelcast/src/main/java/com/hazelcast/map/impl/operation/PutFromLoadAllOperation.java:
```diff
@@ -96,7 +104,7 @@ private void addInvalidation(Data key) {
 
         if (invalidationKeys == null) {
-            invalidationKeys = new ArrayList<>(keyValueSequence.size() / 2);
+            invalidationKeys = new ArrayList<>(loadingSequence.size() / 2);
```
Minor: this is probably very low importance, but shouldn't the initial capacity take into account the presence of `expirationTime` items?

vojtechtoman commented on hazelcast/src/main/java/com/hazelcast/map/EntryLoader.java:
```diff
@@ -0,0 +1,71 @@
+ * @param <K> type of the EntryLoader key
+ * @param <V> type of the EnyryLoader value
+ */
+public interface EntryLoader<K, V> extends MapLoader<K, ExtendedValue<V>> {
```
The Javadoc is more or less an exact copy of the one for `MapLoader`. Do we want this? Wouldn't it better to simply say that `EntryLoader` is an extension of `MapLoader` with support for expiration time? That way, it is immediately clear what the difference between `EntryLoader` and `MapLoader` is, without having to diff the Javadocs of the two.

vojtechtoman commented on hazelcast/src/main/java/com/hazelcast/map/EntryStore.java:
```diff
@@ -0,0 +1,42 @@
+ * @param <K> type of the EntryStore key
+ * @param <V> type of the EntryStore value
+ */
+public interface EntryStore<K, V> extends EntryLoader<K, V>, MapStore<K, ExtendedValue<V>> {
```
The same remark like with `EntryLoader` applies - Javadoc almost identical with `MapStore`.

mustafaiman commented:
@vojtechtoman @ahmetmircik I believe I addressed all comments. Changed ExtendedValue to MetadataAwareValue and moved it into EntryLoader interface as nested class.

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/map/EntryLoader.java:
```diff
@@ -0,0 +1,94 @@
+     * Expiration time is optional.
+     * @param <V> the type of the value object
+     */
+    class MetadataAwareValue<V> {
```
minor: Maybe we can make `MetadataAwareValue` a standalone class since it is also used while storing.

vojtechtoman replied:
:)

mustafaiman replied:
Do you see any technical problem with having it inside EntryLoader? I actually think it make more sense to keep it nested in MapLoader since it is strictly related to it from user's perspective.

ahmetmircik replied:
No technical issue i see, also, since MapStore is also a MapLoader, no big problem from this perspective. But above comment was just my expectation.
